### PR TITLE
Fix ztest mock parameter count issue

### PR DESCRIPTION
**Pull request by jettr**
If you tried to use more mock parameters than was set in the CONFIG_ZTEST_PARAMETER_COUNT, you would get an "unrelated" memory access violation in your test somewhere.

This PR
 - Fixes the out of bound array access
 - Cleans up comments and error strings
 - Cleans up prototype usage of ztest_expect_data to not required caller to cast to `void *`
 - Increases the default `CONFIG_ZTEST_PARAMETER_COUNT` from 1 to 10 when mocking is enabled (still guarded by `CONFIG_ZTEST_MOCKING`)

### CODE REVIEW

**Review by sjg20 — COMMENTED** [review]
commit-message comments

Should use present tense.

Also your first paragraph describes the problem, so should add a sentence explaining what the commit does to fix it.

File: subsys/testsuite/ztest/include/ztest_mock.h [hunk 1]
```diff
@@ -55,33 +55,33 @@
 /**
  * @brief Tell function @a func to expect the data @a data for @a param
  *
- * When using ztest_check_expected_data(), tell that the data ponted by @a param
- * should be same as data at @a data.
- * Only data pointer is stored by this function so it must still be valid when
- * ztest_check_expected_data is called.
+ * When using ztest_check_expected_data(), the data pointed to by
+ * @a param should be same @a data in this function. Only data pointer is stored
```
Comment (sjg20) [inline subsys/testsuite/ztest/include/ztest_mock.h:59]: I think these changes should be in a separate commit

**Review by sjg20 — CHANGES_REQUESTED** [review]
commit-message comments

Should use present tense.

Also your first paragraph describes the problem, so should add a sentence explaining what the commit does to fix it.

File: subsys/testsuite/ztest/include/ztest_mock.h (hunk 1 above)
Comment (sjg20) [reply subsys/testsuite/ztest/include/ztest_mock.h:59]: (I mean your bug fix should be one commit and all your clean-ups in another)

Comment (jettr) [reply subsys/testsuite/ztest/include/ztest_mock.h:59]: Done

**Review by jackrosenthal — APPROVED** [review]

**Comment by jettr** [comment]
From what I can see on the build failures, it looks like sanitycheck just stopped executing without any failures: https://buildkite.com/zephyr/zephyr/builds/4565#5ee1c14a-6653-4a56-b04e-0f7cdb732aea

Is there a way to kick the build again? Is this kind of failure expected/transient? There is still a possibility that there is something wrong with these patches, but they do work for me locally (and there isn't much too them either).

**Review by nashif — APPROVED** [review]

**Review by sjg20 — APPROVED** [review]

**Comment by jettr** [comment]
All checks are passing. @nashif do you merge this or is there some automated process that will merge this since it has all of the necessary approvals and passing checks?
